File: server/rest/model/Entity.php
<?php

/**
 * Created by PhpStorm.
 * User: joeyblack
 * Date: 10/12/16
 * Time: 5:57 PM
 */
abstract class Entity
{

    /**
     * @param string $tableName
     * @param string $statement
     * @return string
     */
    public static function _createDatabaseTableCreateStatementPad($tableName, $statement){
        return "create table if not exists ".$tableName." ("
        .Entity::$database_tableColumn_id." varchar(255) primary key, "
        .$statement
        .Entity::$database_tableColumn_createdAt." float, "
        .Entity::$database_tableColumn_updatedAt." float)";
    }

    public static $database_tableColumn_id = "id";
    public static $database_tableColumn_createdAt = "createdAt";
    public static $database_tableColumn_updatedAt = "updatedAt";

    /*** @var string */
    public $id = Null;

    /*** @var float */
    public $createdAt = Null;

    /*** @var float */
    public $updatedAt = Null;

    /*** @var array */
    public $data;

    /**
     * @return string
     */
    public function getId()
    {
        return $this->id;
    }

    /**
     * @param string $id
     * @return Entity
     */
    public function setId($id)
    {
        $this->id = $id;
        return $this;
    }

    /**
     * @return float
     */
    public function getCreatedAt()
    {
        return $this->createdAt;
    }

    /**
     * @param float $createdAt
     * @return Entity
     */
    public function setCreatedAt($createdAt)
    {
        $this->createdAt = $createdAt;
        return $this;
    }

    /**
     * @return float
     */
    public function getUpdatedAt()
    {
        return $this->updatedAt;
    }

    /**
     * @param float $updatedAt
     * @return Entity
     */
    public function setUpdatedAt($updatedAt)
    {
        $this->updatedAt = $updatedAt;
        return $this;
    }

    /**
     * @return array
     */
    public function getData()
    {
        return $this->data;
    }

    /**
     * @param array $data
     */
    public function setData($data)
    {
        $this->data = $data;
    }



}

File: android/app/src/main/java/ng/edu/aun/tina3/rest/model/Entity.java
package ng.edu.aun.tina3.rest.model;

import com.google.gson.Gson;
import com.google.gson.JsonDeserializationContext;
import com.google.gson.JsonDeserializer;
import com.google.gson.JsonElement;
import com.google.gson.JsonObject;
import com.google.gson.JsonParseException;
import com.google.gson.JsonSerializationContext;
import com.google.gson.JsonSerializer;
import com.google.gson.internal.LinkedTreeMap;

import java.lang.reflect.Type;
import java.util.AbstractMap;
import java.util.HashMap;
import java.util.Map;
import java.util.Set;

import ng.edu.aun.tina3.util.JsonUtils;
import ng.edu.aun.tina3.util.Value;

/**
 * Created by joeyblack on 11/19/16.
 */

public class Entity extends HashMap{

    public static class Constants{
        public static class Fields {
            public static final String CREATED_AT = "createdAt";
            public static final String UPDATED_AT = "updatedAt";
            public static final String DATA = "data";
        }
    }

    private Long createdAt;
    private Long updatedAt;
    private Map<String, String> data;

    public Long getCreatedAt() {
        return createdAt;
    }

    public Entity setCreatedAt(Long createdAt) {
        this.createdAt = createdAt;
        return this;
    }

    public Long getUpdatedAt() {
        return updatedAt;
    }

    public Entity setUpdatedAt(Long updatedAt) {
        this.updatedAt = updatedAt;
        return this;
    }

    public Map<String, String> getData() {
        return data;
    }

    public Entity setData(Map<String, String> data) {
        this.data = data;
        return this;
    }

    /**
     * An abstract Serializer class to handle serializing objects of type {@link Entity} into Json, to be extended
     * registered to a {@link com.google.gson.GsonBuilder} instance before use
     * @see Entity
     * @see com.google.gson.GsonBuilder
     */
    public static abstract class EntitySerializer<T extends Entity> implements JsonSerializer<T> {

        @Override
        public JsonElement serialize(T src, Type typeOfSrc, JsonSerializationContext context) {
            JsonObject jsonObject = new JsonObject();
            serialize(src, jsonObject);
            return performSerialize(src, jsonObject);
        }

        public abstract void serialize(T src, JsonObject jsonObject);

        private JsonElement performSerialize(T t, JsonObject object){
            object.addProperty(Constants.Fields.CREATED_AT, t.getCreatedAt());
            object.addProperty(Constants.Fields.DATA, new Gson().toJson(t.getData()));
            object.addProperty(Constants.Fields.UPDATED_AT, t.getUpdatedAt());
            return object;
        }

    }

    /**
     * An abstract Deserializer class to handle deserializing json into objects of type {@link Entity}, to be extended
     * registered to a {@link com.google.gson.GsonBuilder} instance before use
     * @see Entity
     * @see com.google.gson.GsonBuilder
     */
    public static abstract class EntityDeserializer<T extends Entity> implements JsonDeserializer<T> {
        @Override
        public T deserialize(JsonElement json, Type typeOfT, JsonDeserializationContext context) throws JsonParseException {
            JsonObject o = json.getAsJsonObject();
            return performDeserialize(deserialize(o), o);
        }

        public abstract T deserialize(JsonObject object);

        private T performDeserialize(T t, JsonObject object){
            t.setCreatedAt(Value.TO.longValue(Constants.Fields.CREATED_AT, object));
            t.setUpdatedAt(Value.TO.longValue(Constants.Fields.UPDATED_AT, object));
            try {
                JsonObject o = object.getAsJsonObject(Constants.Fields.DATA);
                if(!Value.IS.nullValue(o)){
                    t.setData(new Gson().fromJson(o, JsonUtils.Data.class));
                }
            }catch (Exception ignored){
                ignored.printStackTrace();
            }

            return t;
        }
    }


}
